### ISSUE TITLE
Symlinks to files outside package root are not being compiled

### ISSUE THREAD

**Issue by PavaniVaka** [ezolenko/rollup-plugin-typescript2#188]
![typescript2 symlink issue](https://user-images.githubusercontent.com/57184116/67865027-461f4580-fb2f-11e9-9e6c-f0836abebc1a.jpg)
When symlink is created to the dependency, rollup is giving error `Error: Unexpected token (Note that you need plugins to import files that are not JavaScript)`

#### Versions
- typescript:^3.6.4
- rollup:"^1.26.0"
- rollup-plugin-typescript2:^0.24.3

**Comment by ezolenko** [comment]
Filesystem symlinks? What OS, soft or hard links? any change you can make a repo with reproduction?

**Comment by PavaniVaka** [comment]
Windows opertating system. Created symlink using "npm link" option to external module. If module is installed then no problem. But to test locally we need to create symlinks and that is not working only with typescript2 plugin. With typescript1 plugin symlinks are working but enums are not working.

**Comment by TerenceZ** [comment]
The same issue, but I use lerna to link packages, and rollup complains unexpected token for the linked packages.

**Comment by thealjey** [comment]
exactly the same issue on windows with lerna
If I copy the folder directly into node_modules everything is fine, but with a symlink it does not work

**Comment by ezolenko** [comment]
@PavaniVaka @TerenceZ @thealjey any chance for a small repo with reproduction?

**Comment by PavaniVaka** [comment]
@ezolenko Sorry, no. But I narrowed down the issue a bit. I noticed that if the external module is copied to application directory then it works. But if it is just symlink then it is not working. For ex: With yalc we can test the external dependencies easily. Because it creates yalc folder in application directory and copies external module to that folder and then creates symlink from yalc folder to node_modules. Hope this gives some clue.

**Comment by thealjey** [comment]
@ezolenko I have actually figured out what the problem was in my case and it has nothing to do with this plugin (it works perfectly) or symlinks for that matter.
It's more of a shortcoming of typescript itself, or rather its configuration relying on json files.
The problem was that the symlink was resolving to a file outside of CWD.
I simply moved my build command to the top level package.json file and everything started working.

**Comment by moki** [comment]
i think we are on the same page here.
#188

> @PavaniVaka @TerenceZ @thealjey any chance for a small repo with reproduction?

@ezolenko https://github.com/moki/mokui there is one.

**Comment by moki** [comment]
I've resolved the issue #194, in case anyone encounters the same problem, below i am providing solution:

When one attempts to build each separate package inside the monorepo, rollup attempts to resolve `@organization/package-name` and include it in the build. You don't want that, so to avoid it upon building each package i am parsing `package.json`, extracting the `dependencies` field's keys, then to check against them inside the `callback` one can provide inside rollup config's `external` field. This will produce the desired outcome.

```
import json from "rollup-plugin-json";

const pkg = process.env.LERNA_PACKAGE_NAME &&
          require(`${process.env.LERNA_PACKAGE_NAME}/package.json`);

const dependencies = ({ dependencies }) => Object.keys(dependencies || {});

const pkgdependencies = dependencies(pkg);

/* exported rollup configuration */
const config = {
    /* your config goes here... */
    /* id is the source name if list of dependencies includes
     * id source name, then mark it as external,
     */
    external: id => pkgdependencies.includes(id)
};

export default config;
```

**Comment by jesse23** [comment]
same issue and `rollup-plugin-typescript` works

**Comment by beenotung** [comment]
The problem also occur when a file under the src is symbolic link to a typescript file outside the project.

e.g. the file structure as below:
```
core/core-client/src/api.ts
myapp/myapp-client/src/api.ts
myapp/myapp-app/src/domain/api.ts -> ../../../myapp-client/src/domain/api.ts
myapp/myapp-app/src/domain/core -> ../../../../core/core-client/src/domain/
...
```

when I compile the project at myapp/myapp-app,
rollup complaint with below error:
```
[ ERROR ]  Rollup: Parse Error: ../myapp-client/src/domain/api.ts:28:7
           Unexpected token (Note that you need plugins to import files that are not JavaScript)

     L28:  export interface IPrimus extends Primus {
```
with the word `interface` highlighted in red

**Comment by beenotung** [comment]
my workaround is to use pre/post hook in npm script to unlink and link the files ...

**Comment by agilgur5** [comment]
> When symlink is created to the dependency, rollup is giving error `Error: Unexpected token (Note that you need plugins to import files that are not JavaScript)`

> * rollup-plugin-typescript2:^0.24.3

> The problem also occur when a file under the src is symbolic link to a typescript file outside the project.

So based on the details mentioned here, this looks to be the same as #216 . While that one came later and describes the problem a bit differently, it also had a reproduction and stated the problem more directly. I was able to root cause both that issue and its rough duplicate, #295 , as well as fix both by upgrading rpt2 to latest.

Per https://github.com/ezolenko/rollup-plugin-typescript2/issues/216#issuecomment-1140024487 and https://github.com/ezolenko/rollup-plugin-typescript2/issues/295#issuecomment-1139994454, this was resolved in rpt2 `0.30.0`, which upgraded [upgraded `@rollup/pluginutils`](https://github.com/ezolenko/rollup-plugin-typescript2/commit/c6f6e52933f520195364758cdf7fe8a371269611) to v4+, which is responsible for filtering files using plugins' `include` / `exclude`.
`pluginutils` v4 released https://github.com/rollup/plugins/pull/517, which removed the `process.cwd()` that it was putting in front of nearly all paths -- the addition of `process.cwd()` upstream made it difficult to `include` files outside of the project root. This could be considered an upstream bug or feature in that respect.


So if you upgrade to rpt2 0.30+ (ideally just use latest), you should be able to transpile files outside of the package root.
(In older versions of rpt2, changing the `include` could have worked, as could changing `rootDir` to `../../`, per the linked comments).
